lib/jsdom/browser/Window.js

```diff
@@ -330,55 +327,151 @@ function Window(options) {
+        }
+      } catch (e) {
+        reportException(window, e, window.location.href);
+      }
+
+      if (repeat && listOfActiveTimers.has(handle)) {
+        timerInitializationSteps(handler, timeout, args, { methodContext, repeat: true, previousHandle: handle });
+      }
+    }
+
+    if (timeout < 0) {
+      timeout = 0;
+    }
+
+    const nodejsTimer = setTimeout(task, timeout);
+    listOfActiveTimers.set(handle, nodejsTimer);
+
+    return handle;
+  }
+
+  // https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#animation-frames
+
+  let animationFrameCallbackId = 0;
+  const mapOfAnimationFrameCallbacks = new Map();
+  let animationFrameNodejsInterval = null;
+
+  // Unlike the spec, where an animation frame happens every 60 Hz regardless, we optimize so that if there are no
+  // requestAnimationFrame() calls outstanding, we don't fire the timer. This helps us track that.
+  let numberOfOngoingAnimationFrameCallbacks = 0;
 
   if (this._pretendToBeVisual) {
-    this.requestAnimationFrame = fn => {
-      const timestamp = rawPerformance.now() - windowInitialized;
-      const fps = 1000 / 60;
-
-      return startTimer(
-        window,
-        setTimeout,
-        clearTimeout,
-        ++latestAnimationFrameCallbackId,
-        fn,
-        fps,
-        animationFrameCallbacks,
-        [timestamp]
-      );
+    this.requestAnimationFrame = function (callback) {
+      callback = webIDLConversions.Function(callback);
+
+      const handle = ++animationFrameCallbackId;
+      mapOfAnimationFrameCallbacks.set(handle, callback);
+
+      ++numberOfOngoingAnimationFrameCallbacks;
+      if (numberOfOngoingAnimationFrameCallbacks === 1) {
+        animationFrameNodejsInterval = setInterval(() => {
+          runAnimationFrameCallbacks(rawPerformance.now() - windowInitialized);
+        }, 1000 / 60);
+      }
+
+      return handle;
     };
-    this.cancelAnimationFrame = stopTimer.bind(this, animationFrameCallbacks);
-  }
 
-  this.__stopAllTimers = function () {
-    stopAllTimers(timers);
-    stopAllTimers(animationFrameCallbacks);
+    this.cancelAnimationFrame = function (handle) {
+      handle = webIDLConversions["unsigned long"](handle);
 
-    latestTimerId = 0;
-    latestAnimationFrameCallbackId = 0;
+      if (mapOfAnimationFrameCallbacks.has(handle)) {
+        --numberOfOngoingAnimationFrameCallbacks;
+        if (numberOfOngoingAnimationFrameCallbacks === 0) {
+          clearInterval(animationFrameNodejsInterval);
+        }
+      }
 
-    timers = Object.create(null);
-    animationFrameCallbacks = Object.create(null);
-  };
+      mapOfAnimationFrameCallbacks.delete(handle);
+    };
+
+    function runAnimationFrameCallbacks(now) {
+      const callbackHandles = mapOfAnimationFrameCallbacks.keys();
+      for (const handle of callbackHandles) {
+        if (mapOfAnimationFrameCallbacks.has(handle)) {
```

Comment:
Can this ever be false? `handle` came from `mapOfAnimationFrameCallbacks` anyway.

It be look simpler to do `for (const [handle, callback] of mapOfAnimationFrameCallbacks) {}` in either case.

Reply:
It can, if the callback calls cancelAnimationFrame. That's why the spec is written as-is. This was fixed in https://github.com/whatwg/html/pull/4375.

Reply:
Ah okay. It also looks like `.delete` doesn’t invalidate existing iterators so this is fine.

Reply:
Hmm, but the iterator skips deleted keys, so maybe this is not necessary?

But, the iterator will continue to get new keys that are added during iteration, which does not match spec behavior... so I think I should snapshot the iterator, and maybe add tests for this...

Reply:
This doesn't seem to have been fixed either. The iterator is indeed "snapshotted", but not its values…